Auto assign listener ports in unit tests

awly opened the pull request:
Use OS-assigned ports (listen on port 0) to avoid manually tracking them via `lib/utils`.
Manual tracking causes conflicts when multiple test processes run in parallel.

As part of this, ad helper methods on `service.TeleportProcess` to expose actual listening addresses of all services.

In local testing, I had no trouble with port exhaustion or sockets stuck in `TIME_WAIT`. I suspect that will happen more in integration tests.

Commit:
Auto-assign ports in utils.LoadBalancer tests

Expose the listening address via new Addr() method.

Commit:
Auto assign listening port in lib/multiplexer tests

Commit:
Auto assign ports for lib/srv/regular tests

Commit:
Add helper methods for listening addresses in TeleportProcess

This is primarily for tests to fetch the actual listening port of these
endpoints when config has port 0.
But it's also convenient for callers to avoid digging into the config
fields.

Commit:
Auto assign ports in lib/service tests

Commit:
Auto assign listening ports in tool/tsh tests

This required splitting combo auth/proxy process into two to get a
random port for auth server that the proxy can join.

Commit:
Auto assign ports in lib/web tests

Also fix a small leak of listeners in test cleanup.

Commit:
Fix data races when accessing internal listeners

These listeners are already protected elsewhere, just missing locking in
a couple methods.

webvictim commented:
There's still a couple of uses of `GetFreeTCPPorts` in the integration tests:

https://github.com/gravitational/teleport/blob/master/integration/integration_test.go#L113
https://github.com/gravitational/teleport/blob/master/integration/kube_integration_test.go#L91

Is the intention to refactor this usage as part of this PR, or were you thinking of doing it separately?

Part of my curiosity with this is because there is also usage of GetFreeTCPPorts in the `teleport-plugins` tests which is causing me a lot of grief (the tests often fail, which I think is partly because ports are being reused)

https://github.com/gravitational/teleport-plugins/blob/master/access/jirabot/jirabot_test.go#L57
https://github.com/gravitational/teleport-plugins/blob/master/access/mattermost/mattermost_test.go#L56
https://github.com/gravitational/teleport-plugins/blob/master/access/pagerduty/pagerduty_test.go#L60
https://github.com/gravitational/teleport-plugins/blob/master/access/slackbot/slackbot_test.go#L57

I was kinda hoping that whatever fix you were mentioning regarding disabling `SO_LINGER` would be a drop-in replacement for `utils.GetFreeTCPPorts` so I could revendor Teleport into `teleport-plugins` and fix the issue. I've probably underestimated the nature of the fix and what actually needs to be done, though.

awly commented:
> There's still a couple of uses of `GetFreeTCPPorts` in the integration tests:
> 
> https://github.com/gravitational/teleport/blob/master/integration/integration_test.go#L113
> https://github.com/gravitational/teleport/blob/master/integration/kube_integration_test.go#L91
> 
> Is the intention to refactor this usage as part of this PR, or were you thinking of doing it separately?

Yes, I'll do the integration tests as a separate PR.
This PR is already big and I anticipate having port exhaustion issues with integration tests.

> 
> Part of my curiosity with this is because there is also usage of GetFreeTCPPorts in the `teleport-plugins` tests which is causing me a lot of grief (the tests often fail, which I think is partly because ports are being reused)
> 
> https://github.com/gravitational/teleport-plugins/blob/master/access/jirabot/jirabot_test.go#L57
> https://github.com/gravitational/teleport-plugins/blob/master/access/mattermost/mattermost_test.go#L56
> https://github.com/gravitational/teleport-plugins/blob/master/access/pagerduty/pagerduty_test.go#L60
> https://github.com/gravitational/teleport-plugins/blob/master/access/slackbot/slackbot_test.go#L57
> 

Right, just to clarify the root cause:
- [GetFreeTCPPorts](https://github.com/gravitational/teleport/blob/master/lib/utils/utils.go#L358-L372) returns sequential ports starting at `20000`
- it's not actually checking that the ports are free
- tests for each Go package run as a standalone test binary
- when running `go test` against multiple packages, it'll run test binaries for multiple packages in parallel
- so we end up with N (number of CPUs) binaries running in parallel, calling `GetFreeTCPPorts`, getting the same exact ports and racing on who gets the listener started first

> I was kinda hoping that whatever fix you were mentioning regarding disabling `SO_LINGER` would be a drop-in replacement for `utils.GetFreeTCPPorts` so I could revendor Teleport into `teleport-plugins` and fix the issue. I've probably underestimated the nature of the fix and what actually needs to be done, though.

Unfortunately it won't be a drop-in replacement.
The separate test processes don't have a way to coordinate on who uses what ports. This means you can't really implement `GetFreeTCPPorts` "correctly" because it doesn't know what ports are free and can't "claim" them.
Only the kernel (or some other external authority) could do this. And the API for that is: "listen on port 0".

There's also an option to use process info and some hashing to do `GetFreeTCPPorts`: e.g. use hash of package name or binary name as the starting port offset. This is hacky, but has much lower chance of collisions than the current implementation.

Looking at the tests you linked, they use <10 ports each. You should be able to refactor to using port 0 without any issues and without needing `SO_LINGER`.
`SO_LINGER` is only needed when you cycle through hundreds of ports quickly.

webvictim commented:
Awesome, thanks for the clarification.

`teleport-plugins` vendors `teleport` and uses `utils.GetFreeTCPPorts` from there which is why a fix in `master` would be super useful. I've written a fix for now which works though - https://github.com/gravitational/teleport-plugins/commit/6653b55fd6051ab61e7d5958e469ff0b36c94c5b

Commit:
Fix a logical race condition in SSH server heartbeat

Heartbeat (announcing metadata about the process to the cluster) has to
start after the actual SSH server starts. Without this, the server will
announce an empty address, since the internal listener isn't populated
yet.

russjones commented on lib/service/listeners.go:
```diff
@@ -0,0 +1,97 @@
+	return process.registeredListenerAddr(listenerProxyTunnel)
+}
+
+func (process *TeleportProcess) registeredListenerAddr(typ listenerType) (*utils.NetAddr, error) {
+	process.Lock()
+	defer process.Unlock()
+	for _, l := range process.registeredListeners {
+		if l.typ == typ {
+			return utils.ParseAddr(l.listener.Addr().String())
+		}
+	}
+	return nil, trace.NotFound("no registered address for type %q", typ)
+}
```
This will be flaky during graceful restart? I wonder if it's better to return an error here if multiple listeners are found? Might even uncover some bugs where things are working with incorrect behavior?

awly replied:
Good point, added a check for multiple listeners per type.
This method is currently only used in tests, so a strict error is good.

Commit:
Check that only one listener per type is found in TeleportProcess

TeleportProcess can have multiple listeners per type during graceful
restart. Return an error from address getters to avoid flaky behavior.
These getters should only get called from tests.
